src/app/forum/store/current-thread/current-thread.reducers.ts

```diff
@@ -0,0 +1,56 @@
+import { createReducer, on } from '@ngrx/store';
+import { websocketClientActions } from '../websocket/websocket.actions';
+import { areSameThreads } from '../../models/threads';
+import { fetchingState, readyState } from 'src/app/utils/state';
+import { fetchThreadInfoActions } from './fetchThreadInfo.actions';
+import { State, initialState } from './current-thread.store';
+import { forumThreadListActions, itemPageActions, threadPanelActions, topBarActions } from './current-thread.actions';
+
+const reducer = createReducer(
+  initialState,
+
+  on(websocketClientActions.eventsReceived, (state, { events }): State => ({
+    ...state,
+    events: readyState(
+      [ ...state.events.data ?? [], ...events ]
+        .sort((a, b) => a.time.valueOf() - b.time.valueOf()) // sort by date ascending
+        .filter((el, i, list) => el.time.valueOf() !== list[i-1]?.time.valueOf()) // remove duplicate (using time as differentiator)
+    )
+  })),
+
+  on(topBarActions.toggleCurrentThreadVisibility, (state): State => ({ ...state, visible: !state.visible })),
+
+  on(forumThreadListActions.showAsCurrentThread, (state, { id }): State => {
+    if (state.id && areSameThreads(state.id, id)) return state;
+    else return {
+      ...state,
+      visible: true,
+      id,
+      info: fetchingState(),
+      events: fetchingState(),
+    };
+  }),
+
+  on(forumThreadListActions.hideCurrentThread, (state): State => ({ ...state, visible: false })),
+
+  on(threadPanelActions.close, (state): State => ({ ...state, visible: false })),
+
+  on(itemPageActions.currentThreadIdChange, (state, { id }): State => {
+    if (state.id && areSameThreads(state.id, id)) return state;
+    else return {
+      ...state,
+      id,
+      info: fetchingState(),
+      events: fetchingState(),
+    };
+  }),
+
+  on(fetchThreadInfoActions.fetchStateChange, (state, { fetchState }): State => {
+    if (!state.id) throw new Error('unexpected: no state id while changing thread info');
```

Comment:
I guess the checks like this could be done before call action, as I don't see a reason to put non valid data to store.

Reply:
It was just an additional guard to "fail fast" in the action is not correctly called. You would just remove that check?

Reply:
This has been moved to an effect as a guard. I am not sure how such a throw in a guard will be handled (while the throw in the effect safely ignore the action), but that's good here.